### PR TITLE
Added Unix Implementations for Element Extracting Functions in OMRSOCK API

### PR DESCRIPTION
**Pull request by caohaley**
- Fixed minor changes to argument name in omrsock_getaddrinfo_length.
- Added Unix Implementations for extracting elements of OMRAddrInfoNode struct.
- Added tests to further test omrsock_getaddrinfo_create_hints and per thread buffer functions,
  using newly implemented element extraction functions.

Issue: #4745

Depends on PR: #4555

### CODE REVIEW

**Comment by caohaley** [comment]
@rwy0717 Could you run the additional tests on the PR?

**Comment by rwy0717** [comment]
@genie-omr build all

File: port/unix/omrsock.c [hunk 1]
```diff
@@ -82,25 +82,50 @@ omrsock_getaddrinfo(struct OMRPortLibrary *portLibrary, char *node, char *servic
...
 
 int32_t
 omrsock_getaddrinfo_family(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, int32_t *family, int32_t index)
 {
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+	omr_os_addrinfo *info = handle->addrInfo;
+	int32_t i = 0;
+
+	for(i = 0; i < index; i++) {
+		info = info->ai_next;
```
Comment (babsingh) [inline port/unix/omrsock.c]: is `info->ai_next` or `info` guaranteed to be non-null? dereferencing a null pointer will cause a seg-fault. there should be  a null check:
```suggestion
		if (info != null) {
		 	 info = info->ai_next;
		}
```
you should also consider error handling in case `info` is null.

File: port/unix/omrsock.c [hunk 2]
```diff
@@ -82,25 +82,50 @@ omrsock_getaddrinfo(struct OMRPortLibrary *portLibrary, char *node, char *servic
...
+	}
+
+	*family = info->ai_family;
```
Comment (babsingh) [inline port/unix/omrsock.c]: null check and error handling?

File: port/unix/omrsock.c [hunk 3]
```diff
@@ -82,25 +82,50 @@ omrsock_getaddrinfo(struct OMRPortLibrary *portLibrary, char *node, char *servic
...
 
 int32_t
 omrsock_getaddrinfo_socktype(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, int32_t *socktype, int32_t index)
 {
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+	omr_os_addrinfo *info = handle->addrInfo;
+	int32_t i = 0;
+
+	for(i = 0; i < index; i++) {
+		info = info->ai_next;
```
Comment (babsingh) [inline port/unix/omrsock.c]: null check and error handling?

File: port/unix/omrsock.c [hunk 4]
```diff
@@ -82,25 +82,50 @@ omrsock_getaddrinfo(struct OMRPortLibrary *portLibrary, char *node, char *servic
...
+	}
+
+	*socktype = info->ai_socktype;
```
Comment (babsingh) [inline port/unix/omrsock.c]: null check and error handling?

File: port/unix/omrsock.c [hunk 5]
```diff
@@ -82,25 +82,50 @@ omrsock_getaddrinfo(struct OMRPortLibrary *portLibrary, char *node, char *servic
...
 
 int32_t
 omrsock_getaddrinfo_protocol(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, int32_t *protocol, int32_t index)
 {
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+	omr_os_addrinfo *info = handle->addrInfo;
+	int32_t i = 0;	
+
+	for(i = 0; i < index; i++) {
+		info = info->ai_next;
```
Comment (babsingh) [inline port/unix/omrsock.c]: null check and error handling?

File: port/unix/omrsock.c [hunk 6]
```diff
@@ -82,25 +82,50 @@ omrsock_getaddrinfo(struct OMRPortLibrary *portLibrary, char *node, char *servic
...
+	}
+
+	*protocol = info->ai_protocol;
```
Comment (babsingh) [inline port/unix/omrsock.c]: null check and error handling?

File: fvtest/porttest/omrsockTest.cpp [hunk 7]
```diff
@@ -104,14 +104,33 @@ TEST(PortSockTest, per_thread_buffer_functionality)
...
+
+	OMRPORTLIB->sock_getaddrinfo_family(OMRPORTLIB, hints, &family, 0);
+	EXPECT_EQ(family, hintsFamily);
+
+	OMRPORTLIB->sock_getaddrinfo_socktype(OMRPORTLIB, hints, &sockType, 0);
+	EXPECT_EQ(sockType, hintsSockType);
+
+	OMRPORTLIB->sock_getaddrinfo_protocol(OMRPORTLIB, hints, &protocol, 0);
+	EXPECT_EQ(protocol, hintsProtocol);
+
```
Comment (babsingh) [inline fvtest/porttest/omrsockTest.cpp]: tests should also cover invalid inputs and verify error handling.

File: port/unix/omrsock.c [hunk 8]
```diff
@@ -82,25 +82,50 @@ omrsock_getaddrinfo(struct OMRPortLibrary *portLibrary, char *node, char *servic
...
+	*socktype = info->ai_socktype;
+	return 0;
 }
 
 int32_t
 omrsock_getaddrinfo_protocol(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, int32_t *protocol, int32_t index)
 {
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+	omr_os_addrinfo *info = handle->addrInfo;
+	int32_t i = 0;
```
Comment (rwy0717) [inline port/unix/omrsock.c]:
```suggestion
	int32_t i = 0;
```

File: port/unix/omrsock.c (hunk 1 above)
Comment (caohaley) [reply port/unix/omrsock.c]: @babsingh Sounds good. I'm going to check if info is null, which would be an error (I can't really check it in the test cases I think because the users can't directly access it). I'm going to also check if handle is NULL, and if index is greater than length, which both would be argument errors (These two will be in the test cases).

For elements of info->ai_next and ai_family and so on, they could be NULL or 0 but they will exist for sure. With the length parameter, as long as index is smaller than length, we shouldn't run into the problem of accessing a NULL pointer.

**Comment by caohaley** [comment]
@babsingh  @rwy0717 Ready for another round of review.

File: port/unix/omrsock.c [hunk 9]
```diff
@@ -80,27 +80,80 @@ omrsock_getaddrinfo(struct OMRPortLibrary *portLibrary, char *node, char *servic
...
+	*length = handle->length;
+	return 0;
 }
 
 int32_t
 omrsock_getaddrinfo_family(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, int32_t *family, int32_t index)
-{
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+{	
+	if (NULL == handle) {
```
[the remark this answers is not in the document]
Comment (rwy0717) [reply port/unix/omrsock.c]: As well, the condition `index > handle->length`, is wrong. The length is an invalid index. You also need to check for negative indices.

File: fvtest/porttest/omrsockTest.cpp [hunk 10]
```diff
@@ -104,14 +105,66 @@ TEST(PortSockTest, per_thread_buffer_functionality)
...
+	rc = OMRPORTLIB->sock_getaddrinfo_protocol(OMRPORTLIB, NULL, &protocol, 0);
+	EXPECT_EQ(rc, OMRPORT_ERROR_INVALID_ARGUMENTS);
+
+	/* Testing invalid arguments: Index is bigger than the length when querying */
+
+	rc = OMRPORTLIB->sock_getaddrinfo_length(OMRPORTLIB, hints, &length);
+	EXPECT_EQ(rc, 0);
+
+	rc = OMRPORTLIB->sock_getaddrinfo_family(OMRPORTLIB, hints, &family, length + 1);
```
Comment (rwy0717) [inline fvtest/porttest/omrsockTest.cpp]: If I understand correctly, the addrinfo list is zero indexed. That means that, if you have three elements, their indices are [0, 1, 2]. The length of the chain, 3, is already "one past the end". There is no need to use length + 1.

You should also test for negative indices.

File: port/unix/omrsock.c [hunk 11]
```diff
@@ -80,27 +80,80 @@ omrsock_getaddrinfo(struct OMRPortLibrary *portLibrary, char *node, char *servic
...
+
+	*family = info->ai_family;
+	return 0;
 }
 
 int32_t
 omrsock_getaddrinfo_socktype(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, int32_t *socktype, int32_t index)
 {
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+	if (NULL == handle) {
```
Comment (rwy0717) [inline port/unix/omrsock.c]: The comments above, also apply here.

File: port/unix/omrsock.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (caohaley) [reply port/unix/omrsock.c]: Yes, you're right sorry. I will fix this.

File: fvtest/porttest/omrsockTest.cpp (hunk 10 above)
Comment (caohaley) [reply fvtest/porttest/omrsockTest.cpp]: You're right, sounds good, I will fix that. Thank you Robert!

File: port/unix/omrsock.c [hunk 12]
```diff
@@ -80,27 +81,71 @@ omrsock_getaddrinfo(struct OMRPortLibrary *portLibrary, char *node, char *servic
...
+	*length = handle->length;
+	return 0;
 }
 
 int32_t
 omrsock_getaddrinfo_family(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, int32_t *family, int32_t index)
-{
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+{	
+	/* Check that the arguments are valid. */
```
Comment (rwy0717) [inline port/unix/omrsock.c]: You probably don't need this comment, it's pretty clear what this code is doing.

**Comment by caohaley** [comment]
@babsingh @rwy0717 I changed the test descriptions and put it into a separate commit. Ready for another round of review.

**Comment by rwy0717** [comment]
@genie-omr build all

**Comment by caohaley** [comment]
@babsingh @rwy0717 Added in the invalid arguments error check, as well as some tests for the error check. Also added in another test to test overwriting of the hints per thread buffer.

Ready for another round of review.

**Comment by rwy0717** [comment]
@genie-omr build all
